klaus/contrib/autodiscover.py

```diff
@@ -0,0 +1,165 @@
+import os
+import dulwich
+import logging
+
+from klaus.repo import FancyRepo
+from klaus import Klaus, make_app
+
+try:
+    from watchdog.observers import Observer
+    from watchdog.events import (
+        FileSystemEventHandler,
+        DirCreatedEvent,
+        DirDeletedEvent,
+    )
+except ImportError:
+    import sys
+    print("To automatically detect and reload repositories as they change, watchdog must be installed")
+    print("see https://pypi.python.org/pypi/watchdog/")
+    sys.exit()
+
+
+logger = logging.getLogger(__file__)
+
+def is_subdirectory(haystack, needle):
+    """
+    checks whether `needle` exists as a subdirectory of `haystack`
+    """
+    def func(arg, dirname, fnames):
+        if dirname == needle or needle in fnames:
+            logger.debug('%s is within %s', needle, haystack)
+            raise StopIteration()
+    try:
+        os.path.walk(haystack, func, None)
+    except StopIteration:
+        return True
+    return False
+
+
+class KlausEventHandler(FileSystemEventHandler):
+    """
+    Basic file-system event handler for a klaus app.
+    When git repositories are created, moved or deleted, watchdog will notify us here.
+    We try and update the klaus wsgi app with the modifications as they happen.
+    """
+    def __init__(self, klausapp, *args, **kwargs):
+        self.app = klausapp
+        return super(KlausEventHandler, self).__init__(*args, **kwargs)
+
+    def _add_dir(self, directory):
+        """
+        Attempts to register `directory` with the klaus app as a git repository.
+        If `directory` is not a git repo, this method does nothing.
+        """
+        try:
+            repo = FancyRepo(directory)
+        except dulwich.errors.NotGitRepository:
+            logger.debug("Dulwich says '%s' is not a git repository", directory)
+            return
+
+        matching = filter(lambda existing_repo: existing_repo.name == repo.name, self.app.repos)
```

Comment:
FYI, I did some more thinking about this, and tried to code around the fact the `repo_map.keys()` are used in the URL to uniquely identify a repository.  I've left it for now.  If you can think of a way to uniquely identify repositories with the same basename that doesn't leak filesystem information in the URL, and keeps the URLs as pretty as they are now, I think this could be revisited, but for now, we'll just have to live with only using repositories that have unique basenames.